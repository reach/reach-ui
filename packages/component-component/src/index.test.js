import React from "react";
import renderer from "react-test-renderer";
import Component from "./index";

let snapshot = element => {
  let wrapper = renderer.create(element);
  const tree = wrapper.toJSON();
  expect(tree).toMatchSnapshot();
  return tree;
};

describe("rendering", () => {
  it("renders without figuritively exploding", () => {
    snapshot(
      <Component>
        <div>Heyyyooooo</div>
      </Component>
    );
  });
  it("prefers `render` over `children`", () => {
    let { children } = snapshot(
      <Component
        initialState={{ value: "!" }}
        render={({ state }) => <div>{state.value}</div>}
      >
        <div>!!</div>
      </Component>
    );
    expect(children).toEqual(["!"]);
  });
  it("renders without children", () => {
    let { root } = renderer.create(<Component initialState={{ value: "!" }} />);
    expect(root.props.initialState.value).toBe("!");
  });
  it("renders with normal children", () => {
    let { root } = renderer.create(
      <Component>
        <div>
          <div value="y">!</div>
        </div>
      </Component>
    );
    expect(root.findByProps({ value: "y" })).toBeTruthy();
  });
  it("renders with children render prop", () => {
    let { children } = snapshot(
      <Component
        initialState={{ value: "!" }}
        render={({ state }) => <div>{state.value}</div>}
      />
    );
    expect(children).toEqual(["!"]);
  });
});

describe.skip("refs", () => {
  it("maintains refs from render to render", () => {
    let wrapper = renderer.create(
      <Component
        getRefs={() => ({ button: React.createRef() })}
        initialState={{ value: "!" }}
        didUpdate={({ refs }) => {
          // refs.button.current.focus()
        }}
        render={({ refs, state, setState }) => (
          <button
            ref={refs.button}
            onClick={() => {
              setState(prevState => ({ value: prevState.value + "!" }));
            }}
          >
            {state.value}
          </button>
        )}
      />
    );
    let tree = wrapper.toJSON();
    expect(tree).toMatchSnapshot();
    tree.props.onClick();
    tree = wrapper.toJSON();
    expect(tree.children).toEqual(["!!"]);
  });
});

describe("state", () => {
  it("receives initialState", () => {
    let wrapper = renderer.create(<Component initialState={{ value: "!" }} />);
    expect(wrapper.toTree().props.initialState).toEqual({ value: "!" });
  });
  it("calls getInitialState", () => {
    let wrapper = renderer.create(
      <Component getInitialState={() => ({ value: "!" })} />
    );
    expect(wrapper.toTree().props.getInitialState()).toEqual({ value: "!" });
  });
  it("updates state", () => {
    let wrapper = renderer.create(
      <Component
        initialState={{ lang: "en" }}
        render={({ state, setState }) => (
          <div
            lang={state.lang}
            onMouseEnter={() => {
              setState({ lang: "fr" });
            }}
          />
        )}
      />
    );
    let tree = wrapper.toJSON();
    expect(tree).toMatchSnapshot();
    tree.props.onMouseEnter();
    tree = wrapper.toJSON();
    expect(wrapper.root.findByProps({ lang: "fr" })).toBeTruthy();
  });
});

// describe("didMount", () => {
//   it("does not require it");
//   it("calls it with the right args");
// });

// describe("willUnmount", () => {
//   it("does not require it");
//   it("calls it with the right args");
// });

// describe("didUpdate", () => {
//   it("does not require it");
//   it("calls it with the right args");
// });

// describe("getSnapshotBeforeUpdate", () => {
//   it("does not require it");
//   it("calls it with the right args");
//   it("returns to cDU correctly");
// });

// describe("shouldUpdate", () => {
//   it("does not require it");
//   it("calls it with the right args");
//   it("returns correctly");
// });
