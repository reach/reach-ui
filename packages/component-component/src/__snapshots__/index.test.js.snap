// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`refs maintains refs from render to render 1`] = `
<button
  onClick={[Function]}
>
  !
</button>
`;

exports[`rendering prefers \`render\` over \`children\` 1`] = `
<div>
  !
</div>
`;

exports[`rendering renders with children render prop 1`] = `
<div>
  !
</div>
`;

exports[`rendering renders without figuritively exploding 1`] = `
<div>
  Heyyyooooo
</div>
`;

exports[`state updates state 1`] = `
<div
  lang="en"
  onMouseEnter={[Function]}
/>
`;
